src/algorithms/crypto/randomness.go

```diff
@@ -0,0 +1,40 @@
+package crypto
+
+import (
+	util "github.com/filecoin-project/specs/util"
+)
+
+type Randomness = util.Randomness
+type Serialization = util.Serialization
+
+// Derive a random byte string from a domain separation tag and an arbitrary
+// serializable object.
+//
+// Note: to produce values of type Serialization, use the auto-generated method prototypes
+//   Serialize_T(T) Serialization
+// for each type T defined in .id.
+//
+// In order to derive randomness from a collection of objects, rather than just a single
+// object, define a struct at the .id level that contains those objects as member fields.
+// This will then cause a Serialize_*() method to be generated for the struct type.
+func (tag DomainSeparationTag) DeriveRand(s Serialization) Randomness {
+	return _deriveRandInternal(tag, s, 0)
+}
+
+// As in DeriveRand(), but additionally accepts an index into the implicit pseudorandom stream.
+// Index must be strictly positive.
+func (tag DomainSeparationTag) DeriveRandExtended(s Serialization, index int) Randomness {
```

Comment:
The method name is not super descriptive. But I really like these interfaces.

Can you add an optional field as well in case we need it (e.g. including minerAddress in ticket production to distinguish tickets generated by different miners with the same key

Reply:
> The method name is not super descriptive.

Good point; how about `DeriveRandWithIndex`?

Reply:
> Can you add an optional field as well in case we need it

I've experimented with variants of this, but I think it would incur too much complexity. If we want to derive randomness from the conjunction of type `A` and optionally type `B`, we can just create a struct `type C = struct { A; B? }` and use the serialization of `C` as the argument here.

Reply:
I disagree only insofar as what the abstraction provides here is inner working of how we form the randomness, e.g. using delimeters, type of encoding, etc. 

If people have to much with that outside the class it defeats part of the purpose imo and remains as error prone as it was before

Reply:
I don't quite understand this point. The definition of `struct C` doesn't need to mention or reason about delimiters or encoding at all; it's just defined the same way as any ordinary struct.

Reply:
fair enough point. yes.